tests: fix HF12 chaingen

ph4r05 opened the pull request:
Currently, the `chaingen.cpp` does not support generating testing block chain with the HF12, which uses RandomX. The problem is that the miner function requires blockchain object to get certain block hashes.

I've created a simple workaround which extracts the block and hardforks from the `events` list. If caller needs to build HF12 blocks, he should set events and network type for the blockchain object. 

It is initial implementation of a working fix, it is not optimal as the blockchain object is constructed with each new block (if HF>=12), but this is robust as there is no need to maintain blockchain db consistency with the tests, it is just created from events when needed.

What do you think?

moneromooo-monero commented:
I'm not sure what you're doing here. The core tests are supposed to use BlockchainDB, or they don't test much at all. They also support v12 blocks (looking now, some of the bulletproof tests are set up to use v12).

ph4r05 commented:
> I'm not sure what you're doing here. The core tests are supposed to use BlockchainDB, or they don't test much at all. They also support v12 blocks (looking now, some of the bulletproof tests are set up to use v12).

I am using `test_generator::construct_block()` to construct a new blockchain for Trezor tests. It creates a testing chain with testing transactions. The generated chain is later tested through the `cryptonote::core` properly. 

The change here just adds a support to generate testing blockchain with the chaingen for HF12. It was not possible before, because there was the code:

```
return cryptonote::get_block_longhash(NULL, b, hash, height, threads);
```

which sets `NULL` as the first parameter (blockchain object). As a result, the nonce computed for a new block with HF12 is computed with invalid RandomX parameters (null hash), as for the genesis block.

With this patch, the chaingen can generate proper RandomX proof-of-work for HF12 blocks.

moneromooo-monero commented:
I've added logs in get_block_longhash, and the core tests do indeed hash with randomx for at least the gen_bp_tx_invalid_bulletproof2_type test.

That said, I've made modifications to the mining code, but I'm 99% sure this test was passing before too.

Using a fake BlockchainDB for the core tests strikes me as wrong. Unit tests can use a fake test db, but core tests are supposed to test the whole thing.

Anyway, I'll post the changes I made to the mining code in case it matters, since I'm pretty much done with it now.

ph4r05 commented:
> Using a fake BlockchainDB for the core tests strikes me as wrong. Unit tests can use a fake test db, but core tests are supposed to test the whole thing.

I believe they do test the whole thing, the only reason the `TestDB` is used in the `test_generator::construct_block()` is to provide `height -> block hash` resolution to the `miner::find_nonce_for_given_block` for HF12 blocks. It may could be done somehow simpler, I don't know. 

The original code:

```cpp
while (!miner::find_nonce_for_given_block([](const cryptonote::block &b, uint64_t height, unsigned int threads, crypto::hash &hash){
    return cryptonote::get_block_longhash(NULL, b, hash, height, threads);
  }, blk, get_test_difficulty(hf_ver), height))
    blk.timestamp++;
```

This leads to the `else` branch for RandomX PoW computation, which should be taken only for genesis block. Thus if caller of the `test_generator::construct_block` specifies he wishes to create HF12 block, this function generates invalid PoW as the Core object which does the full validation has correct blockchain view and the PoW verification takes the ` if (pbc != NULL)` branch, which leads to invalid PoW. 

```cpp
 if (pbc != NULL)
{
  seed_height = rx_seedheight(height);
  hash = pbc->get_pending_block_id_by_height(seed_height);
  main_height = pbc->get_current_blockchain_height();
} else
{
  memset(&hash, 0, sizeof(hash));  // only happens when generating genesis block
  seed_height = 0;
  main_height = 0;
}
```

moneromooo-monero commented:
That does sound right. So your problem is that the hashes are just not what they'd be if they had a different seed hash (ie, the tests work, the block long hashes just aren't equal to what they'd be with a blockchain lookup) ?

ph4r05 commented:
> That does sound right. So your problem is that the hashes are just not what they'd be if they had a different seed hash (ie, the tests work, the block long hashes just aren't equal to what they'd be with a blockchain lookup) ?

Yep exactly. Without this patch the chain produced with chaingen fails the PoW verification for HF12 blocks.

moneromooo-monero commented:
How do you get a failure ?

moneromooo-monero commented:
ie, does gen_bp_tx_invalid_bulletproof2_type work for you or not ?

moneromooo-monero commented:
This is the patch I was talking about:
 https://github.com/monero-project/monero/pull/6111
With that, core tests pass. But it passes an empty seed hash, and the tests still pass. I want to understand why they don't for you first.

moneromooo-monero commented:
I found it. My work branch did not have the "untangle dep" patch yet, and after rebased (and with the patch above), the test I added there now fails. The untangling shouldn't have changed semantics though, but maybe it did for core tests since those are special for the chain.

moneromooo-monero commented:
And do your tests pass if you run them before the untangle commit ?
(I'm not 100% sure that's the commit that started making them fail, I've not bisected yet, but it seems like a very good candidate)

moneromooo-monero commented:
The BP core test still passes though after rebase.

ph4r05 commented:
Sorry for a delay, I got distracted by the weekend.

In fact, I didn't run `gen_bp_tx_invalid_bulletproof2_type` test yet. I've been just using this function in the Trezor functional tests. I am not sure if `gen_bp_tx_invalid_bulletproof2_type` uses `hf_ver` parameter of this function to specify the HF version and wether the core is then later used to verify the block (and the PoW) produced by the mentioned function.

I was just debugging Trezor tests after switch to HF12, I figured out chain verification does not work - it failed with invalid PoW. After few minutes I figured out that the block creating function creates invalid PoW because of the invalid seed hash for HF12 blocks. After I added the `blockchain` object to the miner function the PoW verification passed and Trezor tests run again without error.

I can try also the `gen_bp_tx_invalid_bulletproof2_type` if that helps.

moneromooo-monero commented:
I just want to understand your claim. You claimed tests fail due to bad PoW, but I have at least one test that uses v12 (or v13, but same thing, it's using randomx) and it works. So there is something odd here.

ph4r05 commented:
> I just want to understand your claim. You claimed tests fail due to bad PoW, but I have at least one test that uses v12 (or v13, but same thing, it's using randomx) and it works. So there is something odd here.

I see. I could not find the `gen_bp_tx_invalid_bulletproof2_type` in the master. Where I could find it pls? The question is, whether your test is using this particular function to build the block, with the last parameter set to 12 and whether the Core doing the block verification checks the PoW on the block with the correct seed hash for randomx. I could check that by adding some logging and running your test. In general, I am interested how your test generates PoW for HF13 and passes Core verification with non-zero seedhash.

```cpp
bool test_generator::construct_block(cryptonote::block& blk, uint64_t height, const crypto::hash& prev_id,
                                     const cryptonote::account_base& miner_acc, uint64_t timestamp, uint64_t already_generated_coins,
                                     std::vector<size_t>& block_weights, const std::list<cryptonote::transaction>& tx_list,
                                     const boost::optional<uint8_t>& hf_ver)
```

ph4r05 commented:
Correct me if I am wrong pls, but the 

```cpp
bool gen_bp_tx_validation_base::generate_with()
```

uses the `hf_version` only during the construction of the last block:

- which calls `bool test_generator::construct_block_manually()`,
- this calls the `fill_nonce()`, 
- calls `miner::find_nonce_for_given_block(blk, diffic, height)`
- calls `get_block_longhash(bl, h, height)`
- the `moneromooo/crash` branch does not contain randomX hashing in the `get_block_longhash`

ph4r05 commented:
I've updated the PR, the `fill_nonce` was changed. It uses builds the `blockchain` from `m_events` if HF>=12. If `m_events` are not `NULL` for HF>=12 an exception is thrown.

moneromooo-monero commented:
Sorry, I did not realize this was not on master. I've pushed the current code to the crash branch then. I'm pretty sure I did see randomx logs running this test, and extra logs in get_block_longhash did report the code went in the randomx code path.

ph4r05 commented:
Hm I tried to build it, but there is a problem with the submodule

```
error: Server does not allow request for unadvertised object bc8954ffdb8b2ba8e66982ffa5fe8b8ac0149cca
Fetched in submodule path 'external/randomx', but it did not contain bc8954ffdb8b2ba8e66982ffa5fe8b8ac0149cca. Direct fetching of that commit failed.
```

moneromooo-monero commented:
You can use the upstream randomx, it's just debug-only code I added.

ph4r05 commented:
@moneromooo-monero I've tested the `gen_bp_tx_invalid_bulletproof2_type` and here are my observations:

- PoW is computed *differently* in the block construction and block verification
- `fill_nonce` uses zero seed-hash
- `Blockchain::handle_block_tom_main_chain` uses valid, non-zero seed-hash
- difficulty in `fill_nonce` is 2, but difficulty in the verification is 1 (small chain? small blocks?)
- PoW verification passes even with a random hash when difficulty is 1
- when difficulty raises (as it got raised in my Trezor tests), PoW verification starts to fail

I've added my modifications to my `moneromooo-crash` branch.

Logs from block construction:

```
2019-11-13 12:50:10.301 I BPTEST constructing block
2019-11-13 12:50:10.303 I Filling nonce for h: 73, hf: 14, diffic: 2
2019-11-13 12:50:10.303 I get_long_hash, hf=14, pbc=0
2019-11-13 12:50:10.303 I  .. seed_height: 0, hash: <0000000000000000000000000000000000000000000000000000000000000000>
2019-11-13 12:50:10.329 D Couldn't use largePages for RandomX cache
2019-11-13 12:50:18.780 D Couldn't use largePages for RandomX dataset
2019-11-13 12:50:32.382 D Couldn't use largePages for RandomX VM
2019-11-13 12:50:32.397 I ... <ce1845039d84b16f57e5e626366ae5c4adefffc06a1dcd932b92b744befe8762>
2019-11-13 12:50:32.398 I get_long_hash, hf=14, pbc=0
2019-11-13 12:50:32.398 I  .. seed_height: 0, hash: <0000000000000000000000000000000000000000000000000000000000000000>
2019-11-13 12:50:32.408 I Computed PoW: <ce1845039d84b16f57e5e626366ae5c4adefffc06a1dcd932b92b744befe8762>
```

Logs from verification:

```
2019-11-13 12:50:33.165 I get_long_hash, hf=14, pbc=1
2019-11-13 12:50:33.165 I  .. seed_height: 0, hash: <7cc7044724d5dfa056cb7926f8da0aed28dc4b8c72a6fbbd9ee3008cd7df2f22>
2019-11-13 12:50:41.510 D Couldn't use largePages for RandomX VM
2019-11-13 12:50:41.556 I Cached PoW: <8f02caa23af4d92855bdba4b5e981f9a9f2fe02e1c2a63865c66bab50b79cf8b>
2019-11-13 12:50:41.556 I get_long_hash, hf=14, pbc=1
2019-11-13 12:50:41.556 I  .. seed_height: 0, hash: <7cc7044724d5dfa056cb7926f8da0aed28dc4b8c72a6fbbd9ee3008cd7df2f22>
2019-11-13 12:50:41.579 D Couldn't use largePages for RandomX VM
2019-11-13 12:50:41.621 I Recomputed PoW: <8f02caa23af4d92855bdba4b5e981f9a9f2fe02e1c2a63865c66bab50b79cf8b>
2019-11-13 12:50:41.621 I Validating PoW, hash: <8f02caa23af4d92855bdba4b5e981f9a9f2fe02e1c2a63865c66bab50b79cf8b> at diffic: 1, fixed: 0
2019-11-13 12:50:41.621 I Hash Check: <8f02caa23af4d92855bdba4b5e981f9a9f2fe02e1c2a63865c66bab50b79cf8b>, with diff=2, val: 0
2019-11-13 12:50:41.621 I Random check: <ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff>, val: 1
2019-11-13 12:50:41.621 I Random check: <0000000000000000000000000000000000000000000000000000000000000000>, val: 1
2019-11-13 12:50:41.621 I Random check: <aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa>, val: 1
2019-11-13 12:50:41.621 I Random check: <5555555555555555555555555555555555555555555555555555555555555555>, val: 1
```

As you can see from the `Hash Check` line, the PoW does not pass with designed difficulty 2 (`val: 0`). `Random check` shows, that even those pattern-hashes (PoW) pass the check with difficulty 1.

moneromooo-monero commented:
Oh, so the diff was 1 ? That totally explains then. Thanks.

moneromooo-monero commented on tests/core_tests/chaingen.cpp:
```diff
@@ -55,6 +61,126 @@ using namespace epee;
+  {
+    MDEBUG("Extracting hard-forks from blocks");
+    extract_hard_forks_from_blocks(events, hardforks);
+  }
```
Is there a case where this is needed ? There are always test_options in core_tests (possibly empty, but in all cases the blocks have to match it).

ph4r05 replied:
For the blockchain object construction I need to specify the hard forks so I extract the information from the `events`. In the `init_blockchain` I have no access to `test_options` specified in `core_tests` so I am using this approach.

In the Trezor tests I stored hard fork as a specific event (for some reason). As a fallback for tests not doing that I am reconstructing the hard forks from the blocks stored in the `events`.

moneromooo-monero replied:
OK. Thanks for taking the time to find out and explain.

selsta commented:
@ph4r05 This broke core tests. Can you look into it?

ph4r05 commented:
> @ph4r05 This broke core tests. Can you look into it?

OK sure, I will take a look this week. Can you pls specify which tests broke? It can point me to the right direction before running it all. Thanks!

selsta commented:
```
19:13 <+moneromooo> ph4r05: I've just pulled latest master and some core tests are failing with: "gen_block_big_major_version generation failed: what=events not set, cannot compute valid RandomX PoW". Rings a bell ?
```

Also functional tests fail now on my system but I think that’s unrelated.

ph4r05 commented:
@selsta @moneromooo-monero  Update - I will fix it by Monday.
